File: leetcode/100-999/144. Binary Tree Preorder Traversal/solution_2.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */

/**
 * @param {TreeNode} root
 * @return {number[]}
 */

var preorderTraversal = function(root) 
{
    let tree = []
    let treeVal = []

    while (root || tree.length > 0)
    {
        while (root)
        {
            tree.push(root)
            treeVal.push(root.val)

            root = root.left
        }

        if (tree[tree.length - 1].right)
        {
            const temp = tree[tree.length - 1].right
            tree.pop()
            
            tree.push(temp)
            treeVal.push(temp.val)

            root = temp.left
        }
        else
            tree.pop()
    }    

    return treeVal
};

File: leetcode/100-999/144. Binary Tree Preorder Traversal/solution_2.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution
{
    public:
        vector<int> preorderTraversal(TreeNode *root)
        {
            vector<TreeNode*> tree;
            vector<int> treeVal;

            while (root || !tree.empty())
            {
                while (root)
                {
                    tree.push_back(root);
                    treeVal.push_back(root->val);

                    root = root->left;
                }   

                if (tree.back()->right)
                {
                    auto temp = tree.back()->right;
                    tree.pop_back();

                    tree.push_back(temp);
                    treeVal.push_back(temp->val);
                    
                    root = temp->left;
                }
                else
                {
                    tree.pop_back();
                }
            }

            return treeVal;
        }
};